File: frontend/src/pages/AuditDetail.js
import React, { useEffect, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { getAuditResponse } from '../services/api';
import './shared.css';
import './AuditDetail.css';

const AuditDetail = () => {
  const { id } = useParams();
  const navigate = useNavigate();
  const [response, setResponse] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState('');
  const [selectedImage, setSelectedImage] = useState(null);

  useEffect(() => {
    const loadData = async () => {
      try {
        setLoading(true);
        const data = await getAuditResponse(id);
        setResponse(data);
      } catch (err) {
        setError('Аудитын мэдээлэл ачаалж чадсангүй.');
      } finally {
        setLoading(false);
      }
    };
    loadData();
  }, [id]);

  const getResultBadgeClass = (result) => {
    switch (result) {
      case 'pass':
        return 'badge-success';
      case 'warning':
        return 'badge-warning';
      case 'fail':
        return 'badge-danger';
      default:
        return 'badge-neutral';
    }
  };

  const getResultLabel = (result) => {
    switch (result) {
      case 'pass':
        return 'Амжилттай';
      case 'warning':
        return 'Анхааруулга';
      case 'fail':
        return 'Амжилтгүй';
      default:
        return result || '-';
    }
  };

  if (loading) {
    return (
      <div className="page-container">
        <div className="loading-state">
          <div className="loading-spinner"></div>
          <p>Ачааллаж байна...</p>
        </div>
      </div>
    );
  }

  if (error) {
    return (
      <div className="page-container">
        <div className="error-state">
          <h2>Алдаа гарлаа</h2>
          <p>{error}</p>
          <button className="btn btn-primary" onClick={() => navigate('/audit-results')}>
            Буцах
          </button>
        </div>
      </div>
    );
  }

  if (!response) {
    return (
      <div className="page-container">
        <div className="error-state">
          <h2>Олдсонгүй</h2>
          <p>Аудитын мэдээлэл олдсонгүй.</p>
          <button className="btn btn-primary" onClick={() => navigate('/audit-results')}>
            Буцах
          </button>
        </div>
      </div>
    );
  }

  return (
    <div className="page-container">
      <div className="page-header">
        <div className="page-header-back">
          <button className="btn btn-ghost" onClick={() => navigate('/audit-results')}>
            <span className="back-arrow">&larr;</span> Буцах
          </button>
        </div>
        <div className="page-header-content">
          <h1>Аудитын дэлгэрэнгүй</h1>
          <p className="page-subtitle">ID: {response.id}</p>
        </div>
        <div className="page-actions">
          <span className={`badge ${getResultBadgeClass(response.audit_result)}`}>
            {getResultLabel(response.audit_result)}
          </span>
        </div>
      </div>

      {/* Summary Cards */}
      <section className="detail-summary-grid">
        <div className="content-card summary-card">
          <span className="summary-label">Кампанит ажил</span>
          <span className="summary-value">{response.campaign_name || '-'}</span>
        </div>
        <div className="content-card summary-card">
          <span className="summary-label">Дэлгүүр</span>
          <span className="summary-value">{response.tradeshop_name || '-'}</span>
        </div>
        <div className="content-card summary-card">
          <span className="summary-label">Аудитор</span>
          <span className="summary-value">{response.auditor_name || '-'}</span>
        </div>
        <div className="content-card summary-card">
          <span className="summary-label">Илгээсэн огноо</span>
          <span className="summary-value">{new Date(response.submitted_at).toLocaleString()}</span>
        </div>
      </section>

      {/* Photos Section */}
      <section className="detail-section">
        <div className="section-header">
          <h2>Зургууд</h2>
          <span className="badge badge-neutral">{response.photo_urls?.length || 0}</span>
        </div>
        <div className="photos-grid">
          {(response.photo_urls || []).length === 0 ? (
            <div className="empty-photos">
              <p>Зураг байхгүй байна</p>
            </div>
          ) : (
            response.photo_urls.map((url, index) => (
              <div
                key={index}
                className="photo-card"
                onClick={() => setSelectedImage(url)}
              >
                <img src={url} alt={`Audit photo ${index + 1}`} />
                <div className="photo-overlay">
                  <span>Томруулах</span>
                </div>
              </div>
            ))
          )}
        </div>
      </section>

      {/* Detection Results */}
      <section className="detail-section">
        <div className="section-header">
          <h2>AI Илрүүлэлтийн үр дүн</h2>
          <div className="section-badges">
            <span className="badge badge-active">{response.detection_total || 0} бүтээгдэхүүн</span>
            <span className="badge badge-neutral">{Math.round(response.detection_processing_time_ms || 0)}ms</span>
          </div>
        </div>

        {response.detection_summary && Object.keys(response.detection_summary).length > 0 ? (
          <div className="detection-summary">
            {Object.entries(response.detection_summary).map(([className, count]) => (
              <div key={className} className="detection-item">
                <span className="detection-name">{className}</span>
                <span className="detection-count">{count}</span>
              </div>
            ))}
          </div>
        ) : (
          <div className="empty-detection">
            <p>Илрүүлэлтийн үр дүн алга байна</p>
          </div>
        )}

        {response.detection_items && response.detection_items.length > 0 && (
          <div className="detection-details">
            <h3>Дэлгэрэнгүй илрүүлэлт</h3>
            <div className="detection-list">
              {response.detection_items.map((item, index) => (
                <div key={index} className="detection-detail-item">
                  <span className="detection-class">{item.class_name}</span>
                  <span className="detection-confidence">
                    {Math.round((item.confidence || 0) * 100)}%
                  </span>
                </div>
              ))}
            </div>
          </div>
        )}
      </section>

      {/* Answers Section */}
      <section className="detail-section">
        <div className="section-header">
          <h2>Судалгааны хариултууд</h2>
          <span className="badge badge-neutral">{response.answers?.length || 0}</span>
        </div>

        {(response.answers || []).length === 0 ? (
          <div className="empty-answers">
            <p>Хариулт алга байна</p>
          </div>
        ) : (
          <div className="answers-list">
            {response.answers.map((answer, index) => (
              <div key={answer.question_id || index} className="answer-item">
                <div className="answer-header">
                  <span className="answer-number">Q{index + 1}</span>
                  {answer.auto_answered && (
                    <span className="badge badge-ai">AI</span>
                  )}
                </div>
                <div className="answer-content">
                  <p className="answer-question">{answer.question_text}</p>
                  <p className="answer-value">
                    {Array.isArray(answer.answer)
                      ? answer.answer.join(', ')
                      : String(answer.answer)}
                  </p>
                </div>
              </div>
            ))}
          </div>
        )}
      </section>

      {/* Notes Section */}
      {response.notes && (
        <section className="detail-section">
          <div className="section-header">
            <h2>Тэмдэглэл</h2>
          </div>
          <div className="notes-content">
            <p>{response.notes}</p>
          </div>
        </section>
      )}

      {/* Location Section */}
      {response.location && (
        <section className="detail-section">
          <div className="section-header">
            <h2>Байршил</h2>
          </div>
          <div className="location-content">
            <p>
              <strong>Lat:</strong> {response.location.lat?.toFixed(6)},
              <strong> Lng:</strong> {response.location.lng?.toFixed(6)}
            </p>
            {response.location.accuracy && (
              <p><strong>Нарийвчлал:</strong> {response.location.accuracy.toFixed(1)}м</p>
            )}
          </div>
        </section>
      )}

      {/* Image Modal */}
      {selectedImage && (
        <div className="image-modal-overlay" onClick={() => setSelectedImage(null)}>
          <div className="image-modal-content" onClick={(e) => e.stopPropagation()}>
            <button className="image-modal-close" onClick={() => setSelectedImage(null)}>
              &times;
            </button>
            <img src={selectedImage} alt="Full size" />
          </div>
        </div>
      )}
    </div>
  );
};

export default AuditDetail;

File: application/lib/features/history/models/audit_detail.dart
import 'package:intl/intl.dart';

/// Detection item from YOLO model.
class DetectionItem {
  const DetectionItem({
    required this.className,
    required this.confidence,
    this.bbox,
  });

  final String className;
  final double confidence;
  final List<double>? bbox;

  factory DetectionItem.fromJson(Map<String, dynamic> json) {
    return DetectionItem(
      className: json['class_name']?.toString() ?? 'Unknown',
      confidence: (json['confidence'] as num?)?.toDouble() ?? 0.0,
      bbox: (json['bbox'] as List<dynamic>?)?.cast<num>().map((e) => e.toDouble()).toList(),
    );
  }
}

/// Answer to a survey question.
class AuditAnswer {
  const AuditAnswer({
    required this.questionId,
    required this.questionText,
    required this.answer,
    this.autoAnswered = false,
  });

  final String questionId;
  final String questionText;
  final dynamic answer;
  final bool autoAnswered;

  factory AuditAnswer.fromJson(Map<String, dynamic> json) {
    return AuditAnswer(
      questionId: json['question_id']?.toString() ?? '',
      questionText: json['question_text']?.toString() ?? '',
      answer: json['answer'],
      autoAnswered: json['auto_answered'] == true,
    );
  }

  String get displayAnswer {
    if (answer == null) return '-';
    if (answer is bool) return answer ? 'Тийм' : 'Үгүй';
    return answer.toString();
  }
}

/// Location where audit was performed.
class AuditDetailLocation {
  const AuditDetailLocation({
    required this.lat,
    required this.lng,
    this.accuracy,
  });

  final double lat;
  final double lng;
  final double? accuracy;

  factory AuditDetailLocation.fromJson(Map<String, dynamic> json) {
    return AuditDetailLocation(
      lat: (json['lat'] as num?)?.toDouble() ?? 0.0,
      lng: (json['lng'] as num?)?.toDouble() ?? 0.0,
      accuracy: (json['accuracy'] as num?)?.toDouble(),
    );
  }
}

/// Full audit detail model.
class AuditDetail {
  const AuditDetail({
    required this.id,
    required this.campaignId,
    required this.campaignName,
    required this.tradeshopId,
    required this.tradeshopName,
    required this.auditorId,
    required this.auditorName,
    required this.surveyId,
    required this.answers,
    required this.photos,
    required this.photoUrls,
    required this.status,
    required this.submittedAt,
    required this.createdAt,
    this.detectionId,
    this.detectionSummary,
    this.detectionItems = const [],
    this.detectionTotal = 0,
    this.detectionProcessingTimeMs = 0.0,
    this.location,
    this.notes,
    this.auditResult,
  });

  final String id;
  final String campaignId;
  final String? campaignName;
  final String tradeshopId;
  final String? tradeshopName;
  final String auditorId;
  final String? auditorName;
  final String surveyId;
  final String? detectionId;
  final Map<String, int>? detectionSummary;
  final List<DetectionItem> detectionItems;
  final int detectionTotal;
  final double detectionProcessingTimeMs;
  final List<AuditAnswer> answers;
  final List<String> photos;
  final List<String> photoUrls;
  final AuditDetailLocation? location;
  final String? notes;
  final String status;
  final String? auditResult;
  final DateTime submittedAt;
  final DateTime createdAt;

  String get formattedDate => DateFormat('yyyy-MM-dd HH:mm').format(submittedAt);

  String get resultLabel {
    switch (auditResult) {
      case 'pass':
        return 'Амжилттай';
      case 'warning':
        return 'Анхааруулга';
      case 'fail':
        return 'Амжилтгүй';
      default:
        return auditResult ?? '-';
    }
  }

  factory AuditDetail.fromJson(Map<String, dynamic> json) {
    return AuditDetail(
      id: json['id']?.toString() ?? '',
      campaignId: json['campaign_id']?.toString() ?? '',
      campaignName: json['campaign_name']?.toString(),
      tradeshopId: json['tradeshop_id']?.toString() ?? '',
      tradeshopName: json['tradeshop_name']?.toString(),
      auditorId: json['auditor_id']?.toString() ?? '',
      auditorName: json['auditor_name']?.toString(),
      surveyId: json['survey_id']?.toString() ?? '',
      detectionId: json['detection_id']?.toString(),
      detectionSummary: (json['detection_summary'] as Map<String, dynamic>?)?.map(
        (key, value) => MapEntry(key, (value as num).toInt()),
      ),
      detectionItems: (json['detection_items'] as List<dynamic>?)
              ?.cast<Map<String, dynamic>>()
              .map(DetectionItem.fromJson)
              .toList() ??
          const [],
      detectionTotal: (json['detection_total'] as num?)?.toInt() ?? 0,
      detectionProcessingTimeMs: (json['detection_processing_time_ms'] as num?)?.toDouble() ?? 0.0,
      answers: (json['answers'] as List<dynamic>?)
              ?.cast<Map<String, dynamic>>()
              .map(AuditAnswer.fromJson)
              .toList() ??
          const [],
      photos: (json['photos'] as List<dynamic>?)?.cast<String>() ?? const [],
      photoUrls: (json['photo_urls'] as List<dynamic>?)?.cast<String>() ?? const [],
      location: json['location'] != null
          ? AuditDetailLocation.fromJson(json['location'] as Map<String, dynamic>)
          : null,
      notes: json['notes']?.toString(),
      status: json['status']?.toString() ?? 'completed',
      auditResult: json['audit_result']?.toString(),
      submittedAt: DateTime.tryParse(json['submitted_at']?.toString() ?? '') ?? DateTime.now(),
      createdAt: DateTime.tryParse(json['created_at']?.toString() ?? '') ?? DateTime.now(),
    );
  }
}
